test_examples.sh

```diff
@@ -20,25 +34,86 @@ function cleanup {
 }
 trap cleanup EXIT
 
+export ALL_OPTS=$TEMPDIR/all_opts
+if [[ "$LONG" == "--long" ]]; then
+  p() { [ $# -eq 0 ] && echo || (shift; p "$@") | while read r ; do echo -e "$1,$r\n$r"; done }
+  echo -e $OPTS | p `cat` | sort | uniq | sed 's/,$//' | tail -n +2 > $ALL_OPTS
+else
+  echo "all" > $ALL_OPTS
+fi
+
+
 # Test file $1
 function runtest {
-  echo "running test $1"
-  yes 0 | $SOURIR "$1" --quiet --opt all > $TEMPDIR/$1.opt.out && \
-  yes 0 | $SOURIR "$1" --quiet > $TEMPDIR/$1.out && \
-  diff $TEMPDIR/$1.out $TEMPDIR/$1.opt.out > /dev/null && \
-  yes 1 | $SOURIR "$1" --quiet --opt all > $TEMPDIR/$1.opt.out && \
-  yes 1 | $SOURIR "$1" --quiet > $TEMPDIR/$1.out && \
-  diff $TEMPDIR/$1.out $TEMPDIR/$1.opt.out > /dev/null
-}
+  TEST=$1
+  INPUT=$2
+  OPT=$3
+
+  OUT="$TEMPDIR/$TEST-$INPUT"
+  BASELINE_OUT="$OUT-base.out"
+  OPT_OUT="$OUT-$opt.out"
+
+  # echo "running test $1"
+
+  if [ ! -f $BASELINE_OUT ]; then
+    yes "$INPUT" | $SOURIR "$TEST" --quiet &> "$BASELINE_OUT"
+    if [ $? -ne 0 ]; then
+      echo -e "\n\nBaseline run failed on $TEST with input $INPUT\n"
+      echo " ----- LOG ----------------------------------------"
+      cat $BASELINE_OUT
+      echo " --------------------------------------------------"
+      exit 255
+    fi
+  fi
+
+  yes "$INPUT" | $SOURIR "$TEST" --opt $OPT --quiet &> "$OPT_OUT"
+  if [ $? -ne 0 ]; then
+    echo -e "\n\nOpt run failed on $TEST with input $INPUT and opts $OPT\n"
+    echo " ----- LOG ----------------------------------------"
+    cat $OPT_OUT
+    echo " --------------------------------------------------"
+    exit 255
+  fi
+
+  diff $BASELINE_OUT $OPT_OUT > /dev/null
 
-# Iterate over examples in directory
-ALL_OK=0
-for f in *.sou; do
-  runtest $f
-  if [[ $? -ne 0 ]]; then
-    echo "Example $f failed!"
-    ALL_OK=127
+  if [ $? -ne 0 ]; then
+    echo -e "\n\nOutput differs on $NAME with input $INPUT and opts $OPT\n"
+    echo " ----- DIFF ---------------------------------------"
+    diff $BASELINE_OUT $OPT_OUT
+    echo " --------------------------------------------------"
+    exit 255
   fi
-done
+}
+
+function status {
+  name=$1
+  done=`wc -l < $STATUS`
+  echo -ne "\e[0K\r[${done}/${NUM_TESTS}] ${name} "
```

Comment:
It's fine now that I know what's going on, but the first time, I thought we were running `sum.sou` and `on_without_fallthrough.sou` at the same time, so it was confusing.

I would prefer something clearer and less fancy, e.g. getting rid of the line clearing. In fact, I can live without the counter, especially since the tests (right now) are still very fast.